tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java

```diff
@@ -92,6 +92,9 @@ public LedgerMetadata getLedgerMetadata() {
     @Override
     public CompletableFuture<LedgerEntries> readAsync(long firstEntry, long lastEntry) {
         log.debug("Ledger {}: reading {} - {}", getId(), firstEntry, lastEntry);
+        if (!inputStream.haveDataToRead()) {
```

Comment:
Is this block to be moved below inside the lamdba ?
It looks to me that we are not handling correctly concurrency

Reply:
Yes, I agree this is not thread safe. The calling thread for `readAsync` could be any thread. The `executor` is from an Ordered Executor using the ledger id as the mod. Since all updates to the buffer are run by the same `executor`, the current paradigm is thread safe. We should only update buffer state from within a lambda on this class's executor service.

Reply:
Maybe the method `readAsync` of the class `BlobStoreBackedReadHandleImpl` is not thread-safe, because the seek method will also reset the reader index of the `ByteBuf`.

Reply:
My point is not that `readAsync` is not thread-safe, but rather that updating the buffer's internal state in the way this PR proposes is not thread-safe. I think we should only update the buffer's state from within the lambda to ensure that the buffer is only ever updated by a single thread. Note that the seek method updates the buffer from the executor's single thread in the lambda.

Reply:
exactly, we must move the code inside the lambda

Reply:
Sorry, I misunderstand the problem, nice catch, I'll check the inputStream remaining data in lambda.